fix(build-tools): restore some tsc dep checking

jason-ha opened the pull request:
fix(build-tools): restore some tsc dep checking
fluid-build-tasks-tsc policy handler was searching for command line of exactly "tsc" which has been mostly abandoned since Node16+ESM pattern.

Create a build database that understands basic build tasks, their output, and their CJS/ESM slant. From that, build approximate predecessor lists to replace "tsc" assumption.

The predecessor list is approximate as a package with multiple ESM (or CJS) build scripts may produce the required output from any one of those scripts. Until tsc incremental is run and generates required inputs, precise dependencies cannot be established. This change does not add a post build policy check for such precise dependencies. So, policy addressee must choose the correct predecessor from a list presented. An example of choices presented:
```text
file failed the "fluid-build-tasks-tsc" policy: packages/test/local-server-tests/package.json
'build:test' task is missing the following dependency: 
        @fluid-internal/client-utils#build:test:jest or @fluid-internal/client-utils#build:test:mocha:cjs or @fluid-internal/client-utils#tsc
        @fluid-internal/mocha-test-setup#tsc
        @fluid-private/test-loader-utils#tsc
        @fluid-private/test-pairwise-generator#tsc
        @fluidframework/aqueduct#api-extractor:commonjs or @fluidframework/aqueduct#build:test:cjs or @fluidframework/aqueduct#tsc
        @fluidframework/build-tools#build:test or @fluidframework/build-tools#tsc
        @fluidframework/cell#api-extractor:commonjs or @fluidframework/cell#build:test:cjs or @fluidframework/cell#tsc
```

Relocate GenerateEntrypointsCommand to library/commands (new folder) so database can use shared logic to get output files. Package exports logic was further broken out and placed under test. Supporting changes:
  - promote base.ts to library/commands to prevent import cycle
    - move rootPathFlag from flags.ts directly into base.ts

Apply policy corrections. One of note is api-markdown-documenter that builds ESM from script named tsc.

Follow-up: fluid-build-tasks-eslint should be moved off of getDefaultTscTaskDependencies that looks for "tsc".

jason-ha committed:
fix(build-tools): restore some tsc dep checking
fluid-build-tasks-tsc policy handler was searching for command line of exactly "tsc" which has been mostly abandoned since Node16+ESM pattern.

Create a build database that understands basic build tasks, their output, and their CJS/ESM slant. From that build approximate predecessor lists to replace "tsc" assumption.

The predecessor list is approximate as a package with multiple ESM scripts may produce the required output from any one of those scripts. Until tsc incremental is run and generates required inputs precise dependencies cannot be established. This change does not add a post build policy check for such precise dependencies. So policy addressee must choose the correct predecessor from a list presented.

Refactor logic from entrypoints to be used by database to reason able package exports for `generate entrypoints` scripts.

Apply policy corrections. One of note is api-markdown-documenter that builds ESM from script named tsc.

Future: fluid-build-tasks-eslint should be moved off of getDefaultTscTaskDependencies that looks for "tsc".

tylerbutler commented on build-tools/packages/build-cli/src/library/packageExports.ts:
```diff
@@ -0,0 +1,140 @@
+		throw new Error(`required entrypoints cannot be generated for "exports" array`);
+	}
+
+	// Iterate through exports looking for properties with values matching keys in map.
```
Seems very similar in intent to https://github.com/microsoft/FluidFramework/blob/f44e1f0b950efc624c005545f1ea769d11860535/build-tools/packages/build-cli/src/commands/generate/typetests.ts#L181-L186

jason-ha replied:
I didn't know about `resolve.exports`. I think I might be gathering more information during the processing than it would provide, but I can check it out. I don't think I want to block on getting to that.

jason-ha replied:
Actually, `resolve.exports` doesn't appear helpful. The command this supports works backwards from a check of file path.

tylerbutler commented on build-tools/packages/build-cli/src/library/repoPolicyCheck/fluidBuildDatabase.ts:
It's a little unclear to me when I would use the fluidBuildDatabase vs. the task graph from fluid-build.

jason-ha replied:
Task graph is defined before build and very quick. fluidBuildDatabase is imprecise until after the build completes as it only understands output from commands and not the inputs. So, it can say that one of the tasks in package A is needed by B, but not precisely which of those tasks. After a complete build, tsbuildinfo will have complete inputs required including and ESM/CJS cross over. From that input list we can add a [post-build] policy check on task config that is 100% accurate.

alexvy86 replied:
> After a complete build, tsbuildinfo will have complete inputs required including and ESM/CJS cross over. From that input list we can add a [post-build] policy check on task config that is 100% accurate.

And that input list is what fluidBuildDatabase ends up with, I suppose? Might be worth putting copying this explanation (and a bit more? It's still not 100% clear to me 😅 ) to the class tsdocs.

jason-ha replied:
I mispoke a bit. FluidBuildDatabase doesn't currently end up with that complete information even after build and probably won't. FluidBuildDatabase only has the outputs of scripts. I am adding a comment noting that.
After build we can read the inputs from tsbuildinfo and search FluidBuildDatabase's output knowledge to find source scripts. I don't think we need to store the input into the database, too.

tylerbutler commented on examples/apps/data-object-grid/package.json:
```diff
@@ -110,28 +110,6 @@
 			}
 		}
 	},
-	"fluidBuild": {
-		"tasks": {
-			"eslint": [
-				"^tsc",
-				"@fluid-example/clicker#build:esnext",
-				"@fluid-example/codemirror#build:esnext",
-				"@fluid-example/collaborative-textarea#build:esnext",
-				"@fluid-example/multiview-coordinate-model#build:esnext",
-				"@fluid-example/multiview-slider-coordinate-view#build:esnext",
-				"@fluid-example/prosemirror#build:esnext"
-			],
-			"build:esnext": [
-				"^tsc",
-				"@fluid-example/clicker#build:esnext",
-				"@fluid-example/codemirror#build:esnext",
-				"@fluid-example/collaborative-textarea#build:esnext",
-				"@fluid-example/multiview-coordinate-model#build:esnext",
-				"@fluid-example/multiview-slider-coordinate-view#build:esnext",
-				"@fluid-example/prosemirror#build:esnext"
-			]
-		}
-	},
```
Why are these no longer needed?

jason-ha replied:
They were too limited. There were missing things and now are covered by the base config.
These changes found the hole - this is why we want them.

alexvy86 replied:
I confirmed that these all seem covered by the base config now.

tylerbutler commented on fluidBuild.config.cjs:
```diff
@@ -65,15 +65,10 @@ module.exports = {
+		// Generic build:test script should be replaced by :esm or :cjs specific versions.
+		// "tsc" would be nice to eliminate from here, but plenty of packages still focus
+		// on CommonJS.
+		"build:test": ["typetests:gen", "tsc", "api-extractor:commonjs", "api-extractor:esnext"],
```
This seems like an odd combination. The test script name isn't specific, so if the package builds both commonjs and esm tests, shouldn't both be built by running build:test?

```suggestion
		"build:test": ["build:test:cjs", "build:test:esnext"],
```

jason-ha replied:
If package builds both esm and cjs tests, then those individual test build scripts should have their appropriate dependencies and this isn't needed. This config should be removed, but first the config needs moved to the places that still build tests from build:test. I was doing the minimal changes that are happy with current and new policy and avoiding looking at individual packages.

alexvy86 replied:
> and this isn't needed
I'd say maybe we still want it? We have the single build/compile that covers commonjs and build:esnext, so having the same pattern for tests makes sense to me. That said, for another PR.

tylerbutler commented on packages/service-clients/end-to-end-tests/azure-client/package.json:
```diff
@@ -101,6 +101,14 @@
 		"rimraf": "^4.4.0",
 		"typescript": "~5.1.6"
 	},
+	"fluidBuild": {
+		"tasks": {
+			"build:test": [
+				"^api-extractor:esnext",
+				"^build:esnext"
+			]
+		}
+	},
```
Rather than this, why not update the scripts in this package? Add a build:test:esnext script and make build:test run it. Then you don't need package-specific task overrides.

jason-ha replied:
That would require additional config changes. There is nothing in task config that says test build, that doesn't have a local product build, depends on dependencies packages' product builds. Would be nice if we could comment in package.json easier.

alexvy86 replied:
The package does have a local product build, though, right? `"build": "fluid-build . --task build"`. That said `build:test` indeed does not depend on `build` and thus neither on the dependencies' build output... _But_ `build:test:esm` does seem to enforce that by depending on `build:esnext` which depends on `^build:esnext`. Would an option be to change `build:test` -> `build:test:esm`, change `build` -> `build:esnext` and add a new `build` that just runs `build:esnext`?

alexvy86 replied:
I realized the npm script `build` is defined but there's no tsc script, no tsconfig... so it doesn't do anything locally, just serves for the task dependency graph I guess. Which I had not seen before and I find weird. Not sure we have a good way around it, but pointing it out in case someone has a good idea.

I _think_ the proposal in the previous comments might still work? Not 100% sure.

alexvy86 replied:
And then I saw https://github.com/microsoft/FluidFramework/pull/20971/files#r1594414589, so you can skip my suggestion in this thread.

tylerbutler commented on tools/getkeys/.eslintrc.cjs:
```diff
@@ -9,7 +9,7 @@ module.exports = {
 	],
 	parserOptions: {
-		project: ["./tsconfig.json"],
+		project: ["./tsconfig.lint.json"],
```
Why?

jason-ha replied:
Because it is only used for linting - there is no ts in here and lint config makes packages have a tsconfig, Adding .lint to tsconfig is recognized by policy checks as not a real tsconfig to be used.

tylerbutler commented on tools/api-markdown-documenter/package.json:
```diff
@@ -112,8 +112,8 @@
 		"tasks": {
-			"build:tsc": [
-				"^tsc"
+			"tsc": [
+				"^build:esnext"
```
Why not rename the script to build:esnext instead of tsc?

jason-ha replied:
Because I was not touching scripts and potentially breaking workflows I didn't know about. Leave that for later/others. tools is outside of the fits a standard pattern world and that is not what this task is about.

alexvy86 replied:
I don't see any references to build:tsc in the repo other than the npm script above and the task entry here... not sure what either of them is even accomplishing since nothing else triggers them through `fluid build`. If we change the task entry I think we want to remove the npm script too. @Josmithr might be able to confirm?

jason-ha replied:
We don't need to remove the script entry even if there was a tie in from something else to `build:tsc`. `build:tsc` runs `tsc` which the fluid build system automatically recognizes as a dependency. So, any `build:tsc` task in the system has these manually configured dependencies.
If `build:tsc` script is excessive (unused), then it could be removed.

tylerbutler reviewed (commented):
No objection to these changes, but I wonder if they're needed. What do they protect us against? Would it be easier to use fluid-build's task graph for some of the logic? Seems like you're producing the equivalent here so maybe fluid-build itself could be more involved in finding problems?

jason-ha committed:
merge: 'main' into restore-approximate-build-dependencies

Josmithr commented on build-tools/packages/build-cli/src/library/packageExports.ts:
```diff
@@ -0,0 +1,140 @@
+
+import { ApiTag } from "./apiTag";
+
+export interface ExportData {
```
Docs? 😁

Josmithr commented on build-tools/packages/build-cli/src/library/packageExports.ts:
```diff
@@ -0,0 +1,140 @@
+	return undefined;
+}
+
+export function queryOutputMapsFromPackageExports(
```
Docs? Also the `mapQueryPathToApiTagLevel` param in particular could use some docs explaining the expected key format / what it means for the value to be undefined (with a key presents).

Josmithr commented on tools/getkeys/package.json:
```diff
@@ -24,7 +24,7 @@
 		"start": "node ./index.js",
 		"test": "echo \"Error: no test specified\" && exit 1",
-		"tsc": "tsc"
+		"tsc": "echo \"No build required.\""
```
Can we not just remove this script?

jason-ha replied:
I had no confidence in that. I copied this from changelog-generator-wrapper and expected there was a reason to have it. Will try removal.

Josmithr commented on build-tools/packages/build-cli/src/library/packageExports.ts:
```diff
@@ -0,0 +1,140 @@
+	 */
+	isTypeOnly: boolean;
+}
+export type Node10CompatExportData = Omit<ExportData, "conditions">;
```
Docs explaining this would be good too.

Josmithr commented on build-tools/packages/build-cli/src/library/packageExports.ts:
```diff
@@ -0,0 +1,140 @@
+		if (typeof value === "string") {
+			if (entry === "types") {
+				for (const [key, apiTagLevel] of mapQueryPathToApiTagLevel.entries()) {
+					// eslint-disable-next-line max-depth
```
Nit: extracting the body of the for-loop into its own function might make this disable unecessary.

jason-ha replied:
It would. I felt it was better to read the contents of the function here rather than extract for this aggressive rule (my opinion) and try to name/explain the function and handle the return. ?
We can also mitigate rule violation by making line 47 "if" negative and "continue"ing which would bring this all up one level, but continue can be harder to follow.

Josmithr replied:
I'm not particularly concerned about the line length rule itself, more commenting on the fact that there is a lot of logical nesting going on here. I'm wondering if it might make it a bit more readable to pull out pieces and document what they do in isolation. I'm honestly not sure if it would, but there is a fair amount going on here.

jason-ha replied:
I went in to see if comments would help clear things up and that lead me to the things that aren't handled as we don't author outside of a pattern. I started editing to handle those and found a way to move a little out and I think it is better now.

Josmithr commented on build-tools/packages/build-cli/src/library/repoPolicyCheck/fluidBuildDatabase.ts:
```diff
@@ -0,0 +1,346 @@
+export type PackageNameAndScript = string;
+type AbsoluteFilePath = string;
+type ModuleType = "CommonJS" | "ESM";
+export interface BuildScript {
```
Docs please (plus members). In particular, what does it mean for `moduleType` to be undefined?

alexvy86 commented on packages/test/local-server-tests/package.json:
```diff
@@ -109,6 +109,14 @@
+	"fluidBuild": {
+		"tasks": {
+			"build:test": [
+				"^tsc",
```
For my/reviewer reference, packages have `build:test` depend on `^tsc` + `^api-extractor:commonjs` when they are commonjs and on `^build:esnext` + `^api-extractor:esnext` when they're esm.

Tangent: we should probably make this package esm, but not in this PR.

jason-ha replied:
Yep. And we prefer to have `build:test:cjs|esm` for clarity. `build:test` can exist to run either or both of those but doesn't need dependencies itself. We just need to make sure there are no more `build:test` that do things directly.

Re Tangent: I knew about this. For now, I thought it was good to have something that runs CommonJS. There are some tasks to make sure we have correct coverage on ESM / CJS.

alexvy86 commented on packages/test/test-end-to-end-tests/package.json:
```diff
@@ -149,27 +149,8 @@
-				],
-				"script": false
-			},
-			"build:test": [
-				"...",
-				"^build:esnext"
-			],
```
Unsure if this might still be necessary... unless we update build:test to run build:test:esm as a fluid build task, not just as a npm script, then running build:test as a fluid build task would not ensure dependencies product builds are there, I think?

jason-ha replied:
`build:test` does run `build:test:esm`. fluid build recognizes `npm run <script entry>` and populates the task dependency tree with it.

alexvy86 reviewed (commented):
Haven't looked at the fluidBuildDatabase or the policy nor fluidBuildTasks.ts; the rest looks ok except for potentially one comment left independently above.

jason-ha committed:
tools(tools): remove unused tsc entry

jason-ha committed:
docs(build-tools): comments

jason-ha committed:
improvement(build-tools): make query function out data agnostic

jason-ha committed:
merge: 'main' into restore-approximate-build-dependencies

alexvy86 reviewed (approved):
A few more comments, nothing critical. I can follow the general idea of what's happening and it seems fine, I just can't quite follow the details in my head.

I'll say that several things in these packages (like `hasTaskDependency`) would benefit from unit tests. I'll leave it up to you if it makes sense to add some here or capture work items for the backlog.

jason-ha commented on build-tools/packages/build-cli/src/library/repoPolicyCheck/fluidBuildTasks.ts:
```diff
@@ -351,6 +357,14 @@ function hasTaskDependency(
 	const rootConfig = loadFluidBuildConfig(root);
 	const globalTaskDefinitions = normalizeGlobalTaskDefinitions(rootConfig?.tasks);
 	const taskDefinitions = getTaskDefinitions(json, globalTaskDefinitions, false);
+	// Searched deps that are package specific
+	const packageSpecificSearchDeps = searchDeps.filter((d) => d.includes("#"));
```
Hmm. I think that is less accurate.
1. The package specifier could point to this same package. I don't think they should, but it is possible.
2. Other dependencies might also be from other packages. ^ deps are also from other packages. So, this doesn't capture all of the other package dependencies.
This thing is exactly what it claims. I can add e.g. text though :)

alexvy86 replied:
Fair points :). I don't think we use # for same-package deps in practice even though it seems technically possible (although I wouldn't be super confident the implementation actually works in that case), since we can just use the plain script name. Maybe `Searched task deps that are package specific (we normally use # to indicate dependencies on specific tasks of other packages which are dependencies of the one in question)` or something like that?

jason-ha commented:
> A few more comments, nothing critical. I can follow the general idea of what's happening and it seems fine, I just can't quite follow the details in my head.

Which part is not as followable. This seems like something that should be sufficiently overviewed/explained somewhere (and not just in PR description).

jason-ha commented:
> I'll say that several things in these packages (like `hasTaskDependency`) would benefit from unit tests. I'll leave it up to you if it makes sense to add some here or capture work items for the backlog.

I very much favor having tests for these policies. I generally prefer to avoid whitebox testing but it has its place. Unfortunately, there are no policy tests; so, no infrastructure to add the cases we found are missing. The good news is that we run these against the repo all the time; so, we have a form of testing regularly.
Added AB#8007

jason-ha committed:
docs(build-tools): comments

jason-ha committed:
merge: 'main' into restore-approximate-build-dependencies

Josmithr commented on build-tools/packages/build-cli/src/library/packageExports.ts:
```diff
@@ -0,0 +1,163 @@
+ */
+export function queryOutputMapsFromPackageExports<TOutKey>(
+	packageJson: PackageJson,
+	mapQueryPathToOutKey: Map<string | RegExp, TOutKey | undefined>,
```
Can this be a `ReadonlyMap` since we don't appear to be modifying it here?

jason-ha replied:
Good call

alexvy86 commented:
> Which part is not as followable. This seems like something that should be sufficiently overviewed/explained somewhere (and not just in PR description).

It's mostly lack of familiarity with a good chunk of the code this is touching or calling. Having to go several function calls deep in code that's new to me to actually understand what something in your changes is doing, a few times per file or per function, starts overwhelming my capacity to keep a solid mental model of the system as a whole.

I'd say docs on the helper functions (both newly added ones and pre-existing ones you used) would help, that'd be a good place for further explanation so I could either trust the functions and not have to go through them line by line to understand them, or if I decide to still go through them, at least have some baseline of what I'm expecting the code there to be doing.

Josmithr commented on build-tools/packages/build-cli/src/library/repoPolicyCheck/fluidBuildDatabase.ts:
```diff
@@ -0,0 +1,373 @@
+	pkg: Package,
+	commandLine: string,
+): { files: AbsoluteFilePath[]; type: ModuleType | undefined } | undefined {
+	if (!commandLine.startsWith("flub generate entrypoints")) {
```
Is this check because this function runs as a part of `flub generate entrypoints`? This feels fragile to me. At the very least, it seems like we should extract this into a constant with some docs.

jason-ha replied:
This is not a part of `flub generate entrypoints`. This is helper to derive what output a `flub` command would generate. `generate entrypoints` is the only one I know we need.
It would be better if *Task classes or some part of them could be leveraged. I checked that out briefly but there are too many dependencies and complications.
Same for the fluid commands. Maybe if I understood oclif better there would be a better refactor to reuse the command processing. There is also a complication with code organization and cyclical import dependencies that make relocating Commands hard. Best I could do was extract some common logic. Having said that I might actually be able to clear up the cycle and do something better.

jason-ha replied:
Relocated command and shared more logic

jason-ha committed:
improvement(build-tools): package export proccessing
- tighten up typing with readonly
- some refactoring for readability
- revise isTypeOnly handling to support more exports structuring patterns
- add unit tests for packageExports.ts

jason-ha committed:
merge: 'main' into restore-approximate-build-dependencies

jason-ha committed:
improvement(build-tools): lint and doc updates

jason-ha committed:
improvement(build-tools): tiny tweaks

jason-ha committed:
docs(build-tools): maintenance note

jason-ha committed:
style(build-tools): function rename

jason-ha committed:
refactor(build-tools): relocate GenerateEntrypoints
to library under new commands folder for logic reuse
- promote base.ts to library to prevent import cycle
  - move rootPathFlag from flags.ts

jason-ha committed:
refactor(build-tools): use command logic
to get outputs

jason-ha committed:
refactor(build-tools): dedupe import from refactor

jason-ha committed:
test(build-tools): strip .js from imports
per [ts-node issue783](https://github.com/TypeStrong/ts-node/issues/783)

jason-ha committed:
improvement(client): remove packageExports exports from library
as only use is within library

Josmithr commented on build-tools/packages/build-cli/src/library/commands/generateEntrypoints.ts:
```diff
@@ -0,0 +1,413 @@
+			);
+		}
+
+		await Promise.all(promises);
```
Is it okay to be evaluating `fs.rm` concurrently with the other steps here?

jason-ha replied:
Yes. All of the file reading, writing, and removals are independent. We only read package.json. We only delete or write .d.ts files.

Josmithr replied:
Might be worth a quick note for future readers? But otherwise sounds good then.

Josmithr commented on build-tools/packages/build-cli/src/library/commands/generateEntrypoints.ts:
```diff
@@ -0,0 +1,413 @@
+}
+
+function getOutPathPrefix(
+	{ outDir, outFilePrefix }: { outDir: string; outFilePrefix: string },
```
Nit: It isn't immediately clear to me what `outFilePrefix` is. Is it the file name without extension? Might be worth a comment.

jason-ha committed:
docs(build-tools): added comments

alexvy86 commented on build-tools/packages/build-cli/src/library/packageExports.ts:
```diff
@@ -0,0 +1,214 @@
+ * Read package "exports" to determine which of given file paths are present.
+ *
+ * @param packageJson - json content of package.json
+ * @param mapQueryPathToOutKey - keys of map represent paths to match. When matched
+ * value, if defined, is used to set entry key in output mapKeyToOutput.
```
Took me a couple of reads to convince myself I had (hopefully) understood this one right.
```suggestion
 * @param mapQueryPathToOutKey - keys of map represent paths to match. When one of those paths is
 * found in the package.json's exports, the corresponding value (if defined) is used to set entry key in output mapKeyToOutput.
```

alexvy86 commented on build-tools/packages/build-cli/test/library/packageExports.test.ts:
```diff
@@ -0,0 +1,331 @@
+				types: "./dist/public.d.ts",
+			},
+		},
+		"./alpha": {
+			types: {
+				// a bit unusual authoring where any non-require types lookup
+				// will get resolved. (explicit import condition is common.)
+				require: "./dist/alpha.d.ts",
+				default: "./lib/alpha.d.ts",
+			},
+		},
```
The different structure here compared to all the other test data is to make sure we have a test for it just in case? Or do we actually have a package that does things differently like this?

jason-ha replied:
We do not have any case of this that I know of. This is a legitimate structuring; so, gave it a little coverage.

jason-ha commented on build-tools/packages/build-cli/src/library/commands/generateEntrypoints.ts:
```diff
@@ -0,0 +1,413 @@
+/**
+ * Generates type declarations files for Fluid Framework APIs to support API levels (/alpha, /beta. etc.).
+ */
+export class GenerateEntrypointsCommand extends BaseCommand<
```
Moving just some helpers is what I originally was doing. The real thing that we want to share is the argument processing up to the point of knowing what files are generated. We can't share the argument processing as it is too opaque (at least I couldn't find the way to extract that). So, the next best is to have very similar and somewhat tied together processing in one file and for that I needed the `flags` out of Command. It didn't seem readable to separate `flags` from the Command. So, now they can be in one place and we can access two parts as needed.
Another option was to get repoPolicyCheck out of library, but that appeared more complicated.

jason-ha commented on build-tools/packages/build-cli/test/library/packageExports.test.ts:
```diff
@@ -0,0 +1,331 @@
+	actual: Map<TKeys, TValues>,
+	expected: Map<TKeys, TValues>,
+): void {
+	assert.hasAllKeys(actual, [...expected.keys()]);
```
`assert.hasAllKeys` confirms that keys are identical - those keys and only those keys (I think the name could be better). The below would allow extra keys.

Josmithr replied:
You could simplify all of this as something like 
```suggestion
	assert.deepEquals([...actual.entries()], [...expected.entries()]);
```

I don't know if the error messages would be as useful though.

jason-ha replied:
I liked the output of hasAllKeys. And then there is some nicety to being able to break/debug on single value. Thanks though :)

jason-ha committed:
style(build-tools): rename and doc suggestions

Josmithr commented on build-tools/packages/build-cli/src/library/repoPolicyCheck/fluidBuildDatabase.ts:
```diff
@@ -0,0 +1,369 @@
+			}
+		}
+
+		// Load dependencies outputs too
```
Nit
```suggestion
		// Load dependencies' outputs too
```

Josmithr commented on build-tools/packages/build-cli/src/library/repoPolicyCheck/fluidBuildTasks.ts:
```diff
@@ -351,6 +357,17 @@ function hasTaskDependency(
+	// not a dependency of itself. Skip "name# prefix for self dependencies.
+	const packageSpecificSearchDeps = searchDeps.filter((d) => d.includes("#"));
+	// Set of package dependencies
+	const packageDependencies = new Set([
```
I know we don't currently leverage them, but should this theoretically include `peerDependencies` as well?

jason-ha replied:
Yes, and I think optional peer deps too. This is written to match the task build-tools buildGraph logic that uses Package.combinedDependencies. I'll add a note.

Josmithr commented on build-tools/packages/build-cli/src/library/repoPolicyCheck/fluidBuildTasks.ts:
```diff
@@ -483,7 +520,7 @@ function getTscCommandDependencies(
 	script: string,
 	command: string,
-	defaultDeps: string[],
+	packageMap: Map<string, Package>,
```
Can this be readonly?

Josmithr replied:
Can this be readonly?

Josmithr reviewed (approved):
My overall understanding of how fluid-build functions is pretty limited. But given that, these changes seem reasonable to me.

I would really love to see us invest in moving away from fluid-build more generally, though. The logic involved keeps expanding, and the set of people who know what's going on is very small.

jason-ha committed:
improvement(build-tools): comments and readonly updates

jason-ha committed:
merge: 'main' into restore-approximate-build-dependencies
